packages/astro/src/vite-plugin-astro/hmr.ts

```diff
@@ -55,9 +55,37 @@ const isPkgFile = (id: string | null) => {
 	return id?.startsWith(fileURLToPath(PKG_PREFIX)) || id?.startsWith(PKG_PREFIX.pathname);
 };
 
-export async function handleHotUpdate(ctx: HmrContext, config: AstroConfig, logging: LogOptions) {
-	// Invalidate the compilation cache so it recompiles
-	invalidateCompilation(config, ctx.file);
+export interface HandleHotUpdateOptions {
+	config: AstroConfig;
+	logging: LogOptions;
+	compile: () => ReturnType<typeof cachedCompilation>;
+}
+
+export async function handleHotUpdate(
+	ctx: HmrContext,
+	{ config, logging, compile }: HandleHotUpdateOptions
+) {
+	let isStyleOnlyChange = false;
+	if (ctx.file.endsWith('.astro')) {
+		const oldResult = await compile();
```

Comment:
I assume this is getting the cached version for the given module and therefore not doing 2 compile steps in this function, right?

Reply:
Correct! The first compilation will always hit the cache. I can leave a comment noting this.

Reply:
Comments added in [0c04e21](https://github.com/withastro/astro/pull/4125/commits/0c04e21e81e3f09736666926b2854e8ac68951c7)